hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java

```diff
@@ -2595,8 +2595,15 @@ public OzoneFsServerDefaults getServerDefaults() throws IOException {
 
   @Override
   public URI getKeyProviderUri() throws IOException {
-    return OzoneKMSUtil.getKeyProviderUri(ugi,
-        null, getServerDefaults().getKeyProviderUri(), conf);
+    if (omVersion.compareTo(OzoneManagerVersion.SERVER_DEFAULTS) >= 0) {
+      try {
+        return OzoneKMSUtil.getKeyProviderUri(ugi,
+            null, getServerDefaults().getKeyProviderUri(), conf);
+      } catch (Exception e) {
+        LOG.warn("Could not get key provider URI from OM.", e);
+      }
+    }
```

Comment:
I think it would be better to change `getServerDefaults()` to handle the "old OM" and error cases, in order to simplify life for callers of `getServerDefaults()`.

I'd also consider changing its return type (and `serverDefaults`) to `Optional<OzoneFsServerDefaults>`.

Something like this:

```java
  @Override
  public Optional<OzoneFsServerDefaults> getServerDefaults() throws IOException {
    if (omVersion.compareTo(OzoneManagerVersion.SERVER_DEFAULTS) < 0) {
      return Optional.empty();
    }
    long now = Time.monotonicNow();
    if (!serverDefaults.isPresent() ||
        (now - serverDefaultsLastUpdate > serverDefaultsValidityPeriod)) {
      try {
        serverDefaults = Optional.of(ozoneManagerClient.getServerDefaults());
        serverDefaultsLastUpdate = now;
      } catch (Exception e) {
        LOG.warn("Could not get server defaults from OM.", e);
      }
    }
    return serverDefaults;
  }

  @Override
  public URI getKeyProviderUri() throws IOException {
    String keyProviderUri = serverDefaults.map(FsServerDefaults::getKeyProviderUri)
        .orElse(null);
    return OzoneKMSUtil.getKeyProviderUri(ugi, null, keyProviderUri, conf);
  }
```

Reply:
Thanks for the review @adoroszlai. We are overriding `org.apache.hadoop.fs.FsServerDefaults.getServerDefaults `. The return type is `FsServerDefaults`. Unfortunately `Optional<OzoneFsServerDefaults>` is not compatible with `FsServerDefaults`.
Also, `FsServerDefaults.getKeyProviderUri()` is not a static method. So, I'm are not able to use a method reference
But, I did the error handling to getServerDefaults() and returning null when there is an error.